tests/ignite/handlers/test_ema_handler.py

```diff
@@ -0,0 +1,319 @@
+    model_1.weight.data.fill_(2)
+    state_dict_1 = model_1.state_dict()
+
+    model_2 = _get_dummy_model()
+    ema_handler = EMAHandler(model_2)
+    ema_model = ema_handler.ema_model
+    ema_model.load_state_dict(state_dict_1)
+    torch.testing.assert_allclose(ema_model.weight.data, model_1.weight.data)
+
+
+def test_ema_no_warmup_momentum():
+    data_loader = _get_dummy_dataloader()
+    model = _get_dummy_model()
+    step_fn = _get_dummy_step_fn(model)
+    engine = Engine(step_fn)
+
+    def assert_const_momentum(engine: Engine, const_momentum):
+        assert engine.state.ema_momentum == const_momentum
+
+    # no momentum_warmup
+    ema_handler = EMAHandler(model, momentum=0.002, momentum_warmup=None, warmup_iters=1)
+    ema_handler.attach(engine)
+    # attach the assertion handler after ema_handler, so the momentum is first updated and then tested
+    engine.add_event_handler(Events.ITERATION_COMPLETED, assert_const_momentum, ema_handler.momentum)
+    engine.run(data_loader)
+
+    # no warmup_iters
+    engine = Engine(step_fn)
+    ema_handler = EMAHandler(model, momentum=0.002, momentum_warmup=0.001, warmup_iters=None)
+    ema_handler.attach(engine)
+    # attach the assertion handler after ema_handler, so the momentum is first updated and then tested
+    engine.add_event_handler(Events.ITERATION_COMPLETED, assert_const_momentum, ema_handler.momentum)
+    engine.run(data_loader)
+
+
+def test_ema_update_ema_momentum():
+    data_loader = _get_dummy_dataloader()  # noqa
+    model = _get_dummy_model()
+    step_fn = _get_dummy_step_fn(model)
+    engine = Engine(step_fn)
+
+    warmup_iters = 4
+    momentum_warmup = 0.1
+    momentum = 0.2
+    ema_handler = EMAHandler(model, momentum_warmup=momentum_warmup, momentum=momentum, warmup_iters=warmup_iters)
+    ema_handler.attach(engine)
+
+    # add handlers to check momentum at each iteration
+    @engine.on(Events.ITERATION_COMPLETED)
+    def assert_momentum(engine: Engine):
+        curr_iter = engine.state.iteration
+        curr_momentum = engine.state.ema_momentum
+        if curr_iter == 1:
+            assert curr_momentum == momentum_warmup
+        elif 1 < curr_iter < warmup_iters:
+            assert momentum_warmup < curr_momentum < momentum
+        else:
+            assert curr_momentum == momentum
+
+    engine.run(data_loader, max_epochs=5)
+
+
+def test_ema_buffer():
+    """Test if the tensors in buffer are also synchronized"""
+    model = nn.BatchNorm2d(2)
+    model.running_mean.data.fill_(1.5)
+    model.running_var.data.fill_(1.5)
+    ema_handler = EMAHandler(model)
+
+    def _bn_step_fn(engine, batch):
+        return 1
+
+    engine = Engine(_bn_step_fn)
+    # engine will run 4 iterations
+    engine.run(_get_dummy_dataloader(), max_epochs=2)
+
+    ema_model = ema_handler.ema_model
+    torch.testing.assert_allclose(ema_model.running_mean, model.running_mean)
+    torch.testing.assert_allclose(ema_model.running_var, model.running_var)
```

Comment:
@sandylaker i was inspecting this test and would like to update it to
```python
import torch
import torch.nn as nn

model = nn.BatchNorm2d(2)
model.running_mean.data.fill_(1.5)
model.running_var.data.fill_(1.5)
ema_handler = EMAHandler(model)

def _bn_step_fn(engine, batch):
    x = torch.rand(4, 2, 32, 32)
    _ = model(x)    
    return 1

engine = Engine(_bn_step_fn)
ema_handler.attach(engine)

ema_model = ema_handler.ema_model

@engine.on(Events.ITERATION_COMPLETED)
def check_buffers():
    print(ema_model.running_mean, "vs", model.running_mean)
    torch.testing.assert_allclose(ema_model.running_mean, model.running_mean)
    torch.testing.assert_allclose(ema_model.running_var, model.running_var)

    
# engine will run 4 iterations
engine.run([0, 1], max_epochs=2)
    
torch.testing.assert_allclose(ema_model.running_mean, model.running_mean)
torch.testing.assert_allclose(ema_model.running_var, model.running_var)
```

and it breaks at
```
<ipython-input-5-101d8b59602f> in _update_ema_model(self, engine, name)
    175         momentum = getattr(engine.state, name)
    176         for ema_v, model_v in zip(self.ema_model.state_dict().values(), self.model.state_dict().values()):
--> 177             ema_v.mul_(1.0 - momentum).add_(model_v.data, alpha=momentum)
    178 
    179     def _update_ema_momentum(self, engine: Engine, name: str) -> None:

RuntimeError: result type Float can't be cast to the desired output type Long
```

Reply:
Thanks for checking the test. This error is because the `BatchNorm2d` will create a new buffer `num_batches_tracked` after it forwards input, and the tensor in this buffer has type `long`.